torch/distributed/_spmd/api.py

```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+    ) -> fx.Graph:
+        pass
+
+
+OVERRIDES: Dict[Type[Any], Override] = {}
+
+
+def _register_override(typ: Any, override: Override) -> None:
+    OVERRIDES[typ] = override
+
+
+def _dtensor_expand(gm, args, kwargs, params_and_buffers):
+    flat_args, _ = pytree.tree_flatten(list(args) + list(kwargs.values()))
+
+    mesh = DeviceMesh("cuda", torch.arange(dist.get_world_size()).cuda())
+    shard_schema: Schema = Schema(mesh=mesh, placements=[Shard(0)])
+    replicate_schema: Schema = Schema(mesh=mesh, placements=[Replicate()])
+
+    inps = []
+    schemas = []
+    for a in flat_args:
+        if isinstance(a, torch.Tensor):
+            inps.append(a)
+            schemas.append(shard_schema)
+        elif isinstance(a, nn.Module):
+            # nn.Module placeholder is captured by make_fx but never used in the graph
+            inps.append(torch.empty(0))
+            schemas.append(shard_schema)
+
+    flat_params, _ = pytree.tree_flatten(params_and_buffers)
+
+    for p in flat_params:
+        inps.append(p)
+        schemas.append(replicate_schema)
+
+    # fake_mode: FakeTensorMode = FakeTensorMode()
+    # fake_inps = [fake_mode.from_tensor(inp) for inp in inps]
+
+    return _convert_to_distributed(gm, inps, schemas, _allow_partial=False)[0]
+
+
+def _compile():
+    def inner(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # 1. Extract nn.Module from args and kwargs
+            # FIXME(@mrshenli): detect nn.Module in a nested way
+            # FIXME(@mrshenli): support multiple nn.Module instances
+            # FIXME(@mrshenli): do the same for optimizers
+            mod = None
+            for arg in pytree.tree_flatten(list(args) + list(kwargs.values()))[
+                0
+            ]:
+                if isinstance(arg, nn.Module):
+                    mod = arg
+
+            assert mod is not None
+
+            # 2. Override target submodules (e.g., MoE) with dummy replacements
+            if OVERRIDES:
+                accessor = NamedMemberAccessor(mod)
+
+                for typ, override in OVERRIDES.items():
+                    for name, submodule in mod.named_modules():
+                        if isinstance(submodule, typ):
+                            accessor.swap_submodule(
+                                name, override.replacement(submodule)
+                            )
+
+            # 3. Trace statelss version of the train_step
+            params_and_buffers = {
+                **dict(mod.named_parameters(remove_duplicate=False)),
+                **dict(mod.named_buffers(remove_duplicate=False)),
+            }
+
+            def stateless_func(func, args, kwargs, params_and_buffers):
+                # TODO: do similar thing for optimizer
+                with stateless._reparametrize_module(mod, params_and_buffers):
+                    func(*args, **kwargs)
+                    return [p.grad for p in mod.parameters()]
```

Comment:
I think we need to be faithful to what func returns. Often times it will return the model forward result which is consumed metric/logging related things and we won't capture these in the compiled training step.